File: Templates_Python/DisjointSet.py
# Disjoint Set
class DJS:
    def __init__(self, n):
        self.arr = [i for i in range(n)]
        self.rank = [1] * n
    
    def find(self, x):
        if x != self.arr[x]:
            self.arr[x] = self.find(self.arr[x])
        return self.arr[x]
    
    def union(self, x, y):
        xp = self.find(x)
        yp = self.find(y)
        if xp != yp:
            if self.rank[xp] >= self.rank[yp]:
                self.arr[yp] = xp
                self.rank[xp] += self.rank[yp]
            else:
                self.arr[xp] = yp
                self.rank[yp] += self.rank[xp]
            

File: Templates_cpp/DisjointSet.cpp
#include <tuple>
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <cmath>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <limits>

#define ll long long
#define pii pair<int, int>

using namespace std;

// Disjoint Set
template<typename valueType>
class DJS {
private:
    vector<valueType> arr;
    vector<valueType> rank;

public:
    DJS(valueType len) {
        arr.resize(len);
        for (valueType i = 0; i < len; ++i) arr[i] = i;
        rank.resize(len, 1);
    }

    valueType Find(valueType x) {
        if (arr[x] != x)
            arr[x] = Find(arr[x]);
        return arr[x];
    }

    valueType Union(valueType x, valueType y) {
        valueType xp = Find(x);
        valueType yp = Find(y);
        if (xp != yp) {
            if (rank[xp] >= rank[yp]) {
                arr[yp] = xp;
                rank[xp] += rank[yp];
            } else {
                arr[xp] = yp;
                rank[yp] += rank[xp];
            }
        }
    }
};